utility/fetchClientId.js

```diff
@@ -0,0 +1,100 @@
+/**	Utility function that fetches client IDs (GA3, GA4, Segment, Facebook) from the browser
+ * and sends them asyncrhonously to the Littledata server.
+ * You should run this function when all the tracking libraries have been loaded and when you have
+ * the Shopfy cartToken or checkoutToken created by the Shopify API.
+ *
+ * @param {*} shopifyToken Shopify Cart token or Shopify Checkout token
+ * @param {*} ga3PropertyId GA3 Property ID (ex. UA-XXXX-X)
+ * @param {*} ga4MeasurementId GA4 Measurement ID (ex. G-XXXX)
+ */
+
+const fetchClientId = async ({
+	shopifyToken,
+	ga3PropertyId = undefined,
+	ga4MeasurementId = undefined,
+}) => {
+	const ldPayload = {}
+
+	const getGaIds = async (id, type) => {
+		const getCliendId = new Promise(resolve => {
+			gtag('get', id, type, resolve)
+		})
+		return getCliendId.then(clientId => {
+			switch (type) {
+				case 'client_id':
+					ldPayload['google-clientID'] = clientId
+					break
+				case 'session_id':
+					ldPayload['ga4session-clientID'] = clientId
+					break
+				default:
+					break
+			}
+		})
+	}
+
+	const getSegmentId = async () => {
+		const isSegmentReady = new Promise(resolve => {
+			window.analytics.ready(() => {
+				resolve(window.analytics.user().anonymousId())
+			})
+		})
+		return isSegmentReady.then(id => {
+			ldPayload['segment-clientID'] = id
+		})
+	}
+
+	const getFbId = () => {
+		const getCookie = type =>
+			document.cookie
+				.split('; ')
+				.find(row => row.startsWith(`_${type}=`))
+				?.split('=')[1]
+
+		ldPayload['fbp-clientID'] = getCookie('fbp')
+		ldPayload['fbc-clientID'] = getCookie('fbc')
+	}
+
+	const sendToLittledata = async payload => {
+		const xhr = new XMLHttpRequest()
+		xhr.open(
+			'POST',
+			'https://transactions.littledata.io/v2/clientID/store',
+			true
+		)
+		xhr.setRequestHeader('Content-Type', 'application/json')
+		xhr.send(JSON.stringify(payload))
+	}
+
+	if (shopifyToken) ldPayload.cartID = shopifyToken
```

Comment:
I think this is a must in order to send the client ID. It should not be optional

Reply:
good point